pkg/storage/storage.go

```diff
@@ -122,25 +134,48 @@ func (s *Storage) Directory() string {
 	return s.dir.Path()
 }
 
+// PermanentDatabaseSize returns the size of the underlying permanent database and files.
+func (s *Storage) PermanentDatabaseSize() int64 {
+	return s.permanent.Size()
+}
+
 // PrunableDatabaseSize returns the size of the underlying prunable databases.
 func (s *Storage) PrunableDatabaseSize() int64 {
 	return s.prunable.Size()
 }
 
-// PermanentDatabaseSize returns the size of the underlying permanent database and files.
-func (s *Storage) PermanentDatabaseSize() int64 {
-	return s.permanent.Size()
+// PrunableSQLDatabaseSize returns the size of the underlying prunable SQL database.
+func (s *Storage) PrunableSQLDatabaseSize() int64 {
+	return s.prunableSQL.Size()
 }
 
 func (s *Storage) Size() int64 {
-	return s.PermanentDatabaseSize() + s.PrunableDatabaseSize()
+	return s.PermanentDatabaseSize() + s.PrunableDatabaseSize() + s.PrunableSQLDatabaseSize()
+}
+
+func (s *Storage) RestoreFromDisk() {
+	s.pruningLock.Lock()
+	defer s.pruningLock.Unlock()
+
+	lastPrunedEpoch := s.prunable.RestoreFromDisk()
```

Comment:
So this means we have the lastPrunedEpoch for the prunable and prunable sql store stored in the prunable store. We have to make sure that prunable SQL cannot be in an inconsistent state (i.e. that it retains data that is older than lastPrunedEpoch) when we hook that up with the pruning. Prune on the TransactionRetainer is currently not called outside of tests, so I guess that is yet to be done.

Reply:
I hooked to the `Pruned` event of the storage now, to trigger pruning inside the txRetainer.
There can be some inconsistencies, in case the pruning of the storage stops mid-process, but I think that's fine for now. It will fix itself with the next successful pruning. So basically it's the same inconsistency as if you call the different API endpoints during an ongoing pruning process.